master/buildbot/secrets/providers/vault_hvac.py

```diff
@@ -0,0 +1,164 @@
+# This file is part of Buildbot.  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+"""
+HVAC based providers
+"""
+
+
+from buildbot import config
+from buildbot.secrets.providers.base import SecretProviderBase
+
+import hvac
+
+
+class HvacAuthenticator:
+    """
+    base HVAC authenticator class
+    """
+
+    def authenticate(self, client):
+        pass
+
+
+class HvacAuthenticatorToken(HvacAuthenticator):
+    """
+    HVAC authenticator for static token
+    """
+
+    def __init__(self, token):
+        self.token = token
+
+    def authenticate(self, client):
+        client.token = self.token
+
+
+class HvacAuthenticatorApprole(HvacAuthenticator):
+    """
+    HVAC authenticator for Approle login method
+    """
+
+    def __init__(self, roleId, secretId):
+        self.roleId = roleId
+        self.secretId = secretId
+
+    def authenticate(self, client):
+        client.auth.approle.login(role_id=self.roleId, secret_id=self.secretId)
+
+
+class HvacKvSecretProvider(SecretProviderBase):
+    """
+    basic provider where each secret is stored in Vault KV secret engine
+    """
+
+    name = 'SecretInVaultKv'
+
+    def checkConfig(self, vaultServer=None, authenticator=None, secretsmount=None,
```

Comment:
Prefer snake_case for new code.

Reply:
snake_case for all of: method names, method parametres, variables and class attributes?

Reply:
Yes, except of course when you're overriding existing method, as is `checkConfig` in this case :)

Reply:
Could we just support the old API as well? I think it would make more sense to just keep the old name :)

You can support the old API via kwargs, and just adapt it to the new api. Those are not to far away, you just have to create an authenticator based on the token (and warn_deprecated)

Reply:
There already was discussion regarding this topic in https://github.com/buildbot/buildbot/issues/5903.

Honestly, old implementation has so many flaws, that it is hardly usable in real world scenarios. Keeping backwards compatibility would bloat the code with heap of confusing conditionals and bug simulations to support just few experimental installations.

In this particular case, I really do prefer keeping the code clean and just deprecate old implementation.

Reply:
@tardyp There are edge cases which make migration harder than needed (a mandatory extra argument for all users). So I agree that a separate implementation makes sense here.

Reply:
I agree with your arguments.

What about calling the new one HashiCorpVaultSecretProvider and the older one LegacyHashiCorpVaultSecretProvider?

We can provide detection of use of the older api, and the very few users can either  use LegacyHashiCorpVaultSecretProvider or upgrade the few secrets they have to use new api.

It annoys me to need to forever choose a bad name to keep upgrade seemless for very few user.

https://martinfowler.com/bliki/TwoHardThings.html

Reply:
The problem is that we can't detect the use of the old API reliably. Though maybe we can just use differently named arguments just for the sake of upgrade process.

Reply:
I think we can just add (vaultToken=None) argument, and if this is not None, then this is a usage of the old version of the API. 
Then the user would have the responsability to update its secret renders

Reply:
Well I suggest `HashiCorpVaultKvSecretProvider` then, with added "Kv" part (basically substitute `s/Hvac/HashiCorpVault/` in my original name). 

Why I did add "Kv" to name? When I dug deeper in hvac implementation I realized it would be very complicated to have support for multiple secret engines in single class, as they can be very different and on both hvac and in vault they have completely separate independent API. So I decided to follow vault and hvac pattern in this regard and implement kv secret engine as separate class. In case more secret engines will be implemented in future, each will have it's own class (authentication and common functionalities can be moved to some base class when necessary).

Reply:
> I think we can just add (vaultToken=None) argument, and if this is not None, then this is a usage of the old version of the API.
> Then the user would have the responsability to update its secret renders

Well, for API to be really compatible I would also have to somehow handle default value for `apiVersion`, which is different between the two. While it is possible, it would also bloat the code (and even duplicated thanks to `reconfigService()` vs. `checkConfig()`). And all it would help would be slightly simpler change in secretProvider configuration? User still have to rewrtite all secret identifiers, so he has to invest significantly more effort to migrate to new vault provider anyway. And not to mention that token login method is not suitable for production use (because of enforced TTL on vault tokens), so in the end all it would help are some experimental installations. Is it really worth it?

Reply:
@tardyp My vote would be to just have a separate class. The current implementation is quite broken and the user will need to put quite a lot to migrate anyway. Adding a requirement to rename a class would be a minor inconvenience.

Reply:
I am okay with the HashiCorpVaultKvSecretProvider.

It makes sense as it only supports KV indeed and not other secrets. We need to be very clear in the doc about the difference as it might be a bit too subtle.

Reply:
OK, I'll update docs accordingly.